src/guide/typescript-support.md

```diff
@@ -0,0 +1,270 @@
+
+[WebStorm](https://www.jetbrains.com/webstorm/) also provides out-of-the-box support for both TypeScript and Vue.
+
+## Defining Vue components
+
+To let TypeScript properly infer types inside Vue component options, you need to define components with `defineComponent` global method:
+
+```ts
+import { defineComponent } from 'vue'
+
+const Component = defineComponent({
+  // type inference enabled
+})
+```
+
+## Using with Options API
+
+You can type your `data` properties using type assertions:
+
+```js
+const Component = defineComponent({
+  data() {
+    return {
+      message: 'Hello!' as string,
+      count: 0 as number
+    }
+  },
+  mounted() {
+    const result = this.count.split('') // => Property 'split' does not exist on type 'number'
+  }
+})
+```
+
+If you need a complex type, you can create an interface:
+
+```ts
+interface Book {
+  title: string
+  author: string
+  year: number
+}
+
+const Component = defineComponent({
+  data() {
+    return {
+      message: 'Hello!' as string,
+      count: 0 as number,
+      book: <Book>{
+        title: 'Vue 3 Guide',
+        author: 'Vue Team',
+        year: 2020
+      }
+    }
+  }
+})
+```
+
+### Annotating Return Types
+
+Because of the circular nature of Vue’s declaration files, TypeScript may have difficulties inferring the types of certain methods. For this reason, you may need to annotate the return type on methods like render and those in computed.
+
+```ts
+import { defineComponent } from 'vue'
+
+const Component = defineComponent({
+  data() {
+    return {
+      message: 'Hello!' as string
+    }
+  },
+  methods: {
+    // need annotation due to `this` in return type
+    greet(): string {
+      return this.message + ' world'
+    }
+  },
+  computed: {
+    // need annotation
+    greeting(): string {
+      return this.message + '!'
+    }
+  }
+})
+```
+
+### Annotating Props
+
+```ts
+{ defineComponent, PropType } from 'vue'
+
+interface ComplexMessage {
+  title: string
+  okMessage: string
+  cancelMessage: string
+}
+const Component = defineComponent({
+  props: {
+    name: String,
+    success: { type: String },
+    callback: {
+      type: Function as PropType<() => void>
+    },
+    message: {
+      type: Object as PropType<ComplexMessage>,
+      required: true,
+      validator(message: ComplexMessage) {
+        return !!message.title
+      }
+    }
+  }
+})
+```
+
+If you find validator not getting type inference or member completion isn’t working, annotating the argument with the expected type may help address these problems.
+
+## Using with Composition API
+
+On `setup()` function, you don't need ass a typing to `props` parameter as it will infer types from `props` component option.
+
+```ts
+import { defineComponent } from 'vue'
+
+const Component = defineComponent({
+  props: {
+    message: {
+      type: String,
+      required: true
+    }
+  },
+
+  setup(props) {
+    /* props will be typed as
+    (parameter) props: Readonly<{
```

Comment:
No sure about providing the full type here, probably just showing the `props.message.split('')` is valid because is string